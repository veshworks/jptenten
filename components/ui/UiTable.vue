<script setup lang="ts">
type HeadDef = { text: string; align: 'start' | 'center' | 'end' };
const props = defineProps<{
  title?: string;
  headers?: (string | HeadDef)[];
  items?: string[][];
}>();

const headers = computed(() => {
  return (
    props.headers?.map((item) => {
      if (typeof item === 'string') {
        return {
          text: item,
          align: 'center',
        } as HeadDef;
      }

      return item;
    }) ?? []
  );
});
</script>

<template>
  <table
    class="ui-table"
    cellspacing="0"
    cellpadding="0"
  >
    <caption v-if="title">
      {{
        title
      }}
    </caption>
    <thead>
      <tr>
        <th
          v-for="head in headers"
          :key="head.text"
          :style="{ 'text-align': head.align }"
        >
          {{ head.text }}
        </th>
      </tr>
    </thead>
    <tbody>
      <tr
        v-for="row in items"
        :key="row.join()"
      >
        <td
          v-for="(cell, column) in row"
          :key="cell"
          :style="{ 'text-align': headers[column].align }"
        >
          {{ cell }}
        </td>
      </tr>
    </tbody>
  </table>
</template>

<style lang="scss">
.ui-table {
  font-family: var(--font-family-sans);

  & > caption {
    color: var(--color-red-deep);
    font-family: var(--font-family-attention);
    font-size: var(--font-size-300);
  }

  /* heading and cell */
  & > thead > tr > th,
  & > tbody > tr > td {
    padding-left: var(--space-0100);
    padding-right: var(--space-0100);
  }

  & > thead > tr,
  & > tbody > tr:nth-child(2n) {
    background-color: var(--color-beige-deep);
  }

  /* heading */
  & > thead > tr > th {
    font-family: var(--font-family-attention);
    font-size: var(--font-attention-size);
    text-align: start;
    vertical-align: bottom;
  }

  /* cell */
  & > tbody > tr > td {
    text-align: center;
  }
}
</style>
