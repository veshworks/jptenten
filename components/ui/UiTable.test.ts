// @vitest-environment nuxt
import { describe, it, expect, afterEach } from 'vitest';
import UiTable from './UiTable.vue';
import { render, screen } from '@testing-library/vue';
import { lazylet } from '@vhoyer/lazy-let';

describe('ui components > ui-table', () => {
  const $0 = lazylet(afterEach, {
    props: () => ({}),
    wrapper: $ => render(UiTable, {
      props: $.props,
    }),
    screen: ($) => {
      $.wrapper;
      return screen;
    },
  });

  it('renders empty', () => {
    expect($0.wrapper.container).toMatchSnapshot();
  });

  describe('when a title, headers, and items are passed', () => {
    const $1 = $0({
      props: () => ({
        ...$0.props,
        title: 'Distribuição de trabalho',
        headers: ['nome', 'tarefa'],
        items: [
          ['fulano', 'louça'],
          ['fulano', 'aspirar'],
          ['fulano', 'lavar roupas'],
          ['hoyer', 'assistir youtube'],
        ],
      }),
    });

    it('displays the title', () => {
      expect($1.wrapper.container).toMatchSnapshot();
    });

    describe('when the header defines an alignment', () => {
      const $2 = $1({
        props: () => ({
          ...$1.props,
          headers: ['nome', { text: 'tarefa', align: 'left' }],
        }),
      });

      it('aligns the second column to the left', () => {
        expect($2.wrapper.container).toMatchSnapshot();
      });
    });
  });
});
