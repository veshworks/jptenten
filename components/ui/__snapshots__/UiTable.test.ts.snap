// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ui components > ui-table > renders empty 1`] = `
<div>
  <table
    cellpadding="0"
    cellspacing="0"
    class="ui-table"
  >
    <!--v-if-->
    <thead>
      <tr>
        
        
      </tr>
    </thead>
    <tbody>
      
      
    </tbody>
  </table>
</div>
`;

exports[`ui components > ui-table > when a title, headers, and items are passed > displays the title 1`] = `
<div>
  <table
    cellpadding="0"
    cellspacing="0"
    class="ui-table"
  >
    <caption>
      Distribuição de trabalho
    </caption>
    <thead>
      <tr>
        
        <th
          style="text-align: center;"
        >
          nome
        </th>
        <th
          style="text-align: center;"
        >
          tarefa
        </th>
        
      </tr>
    </thead>
    <tbody>
      
      <tr>
        
        <td
          style="text-align: center;"
        >
          fulano
        </td>
        <td
          style="text-align: center;"
        >
          louça
        </td>
        
      </tr>
      <tr>
        
        <td
          style="text-align: center;"
        >
          fulano
        </td>
        <td
          style="text-align: center;"
        >
          aspirar
        </td>
        
      </tr>
      <tr>
        
        <td
          style="text-align: center;"
        >
          fulano
        </td>
        <td
          style="text-align: center;"
        >
          lavar roupas
        </td>
        
      </tr>
      <tr>
        
        <td
          style="text-align: center;"
        >
          hoyer
        </td>
        <td
          style="text-align: center;"
        >
          assistir youtube
        </td>
        
      </tr>
      
    </tbody>
  </table>
</div>
`;

exports[`ui components > ui-table > when a title, headers, and items are passed > when the header defines an alignment > aligns the second column to the left 1`] = `
<div>
  <table
    cellpadding="0"
    cellspacing="0"
    class="ui-table"
  >
    <caption>
      Distribuição de trabalho
    </caption>
    <thead>
      <tr>
        
        <th
          style="text-align: center;"
        >
          nome
        </th>
        <th
          style="text-align: left;"
        >
          tarefa
        </th>
        
      </tr>
    </thead>
    <tbody>
      
      <tr>
        
        <td
          style="text-align: center;"
        >
          fulano
        </td>
        <td
          style="text-align: left;"
        >
          louça
        </td>
        
      </tr>
      <tr>
        
        <td
          style="text-align: center;"
        >
          fulano
        </td>
        <td
          style="text-align: left;"
        >
          aspirar
        </td>
        
      </tr>
      <tr>
        
        <td
          style="text-align: center;"
        >
          fulano
        </td>
        <td
          style="text-align: left;"
        >
          lavar roupas
        </td>
        
      </tr>
      <tr>
        
        <td
          style="text-align: center;"
        >
          hoyer
        </td>
        <td
          style="text-align: left;"
        >
          assistir youtube
        </td>
        
      </tr>
      
    </tbody>
  </table>
</div>
`;
